lib/Analysis/Allocation.cpp

```diff
@@ -161,9 +161,16 @@ class AllocationAnalysis {
       // TODO(Keren): Reduce with index is not supported yet.
       auto value = op->getOperand(0);
       if (auto tensorType = value.getType().dyn_cast<RankedTensorType>()) {
+        auto srcLayout = tensorType.getEncoding();
+        bool fastReduce = reduceOp.axis() == getOrder(srcLayout)[0];
         auto smemShape = getScratchConfigForReduce(reduceOp);
         unsigned elems = std::accumulate(smemShape.begin(), smemShape.end(), 1,
                                          std::multiplies{});
+        if (fastReduce) {
+          auto mod = op->getParentOfType<ModuleOp>();
+          unsigned numWarps = triton::gpu::TritonGPUDialect::getNumWarps(mod);
+          elems = std::max<unsigned>(elems, numWarps * 32);
```

Comment:
Related link: https://github.com/openai/triton/blob/triton-mlir/lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp#L1753

The required shared memory size is `max(getScratchConfigReduce, numOfThreads)`.